sympy/core/relational.py

```diff
@@ -513,7 +513,7 @@ def expand(self, **kwargs):
         return self.func(*args)
 
     def __bool__(self):
-        raise TypeError("cannot determine truth value of Relational")
+        raise TypeError(f"cannot determine truth value of Relational {self}")
```

Comment:
Generally, string printing can be very expensive operation and it is not encouraged to format the error messages with parameters because the evaluation happens in the order as:

- `f"cannot determine truth value of Relational {self}"` is computed
- TypeError is raised

So if someone is using the SymPy inside try-catch block, it will generally degrade the performance of that.

Anyway, I think that every error messages should be static (no parameters)
Usually such formatting like this is not our job, 
and is not very helpful because it is usually a job for debuggers.

Reply:
Something like below should work to reduce the overhead.

```python
class LazyExceptionMessage:
  def __init__(self, msg, *args, **kwargs):
    self.msg = msg
    self.args = args
    self.kwargs = kwargs

  def __str__(self):
    return self.msg.format(*self.args, **self.kwargs)

raise TypeError(LazyExceptionMessage("cannot determine truth value of Relational {}", "123"))
```

Reply:
If y'all think LazyExceptionMessage is a good fix for this problem, I'm happy to do that

Reply:
I have different suggestion for `LazyExceptionMessage` that I can give some clues why it is best:

```python
class LazyExceptionMessage:
    callback: Callable[[], str]

    def __init__(self, callback): 
        self.callback = callback

    def __str__(self):
        return self.callback()
```
```python
raise TypeError(LazyExceptionMessage(lambda: f"cannot determine truth value of Relational {self}"))
```

- It is by far the most general implementation of lazy strings that works with every formatting styles of strings (`%` formatting, `format`, f-strings). It generally sucks if I only stick to one sort of string formatting, and does not allow other sort of formatting. (For example, f-strings are most encouraged nowadays)
- It is easiest to use, you can just define simple function with `def` or `lambda`. You generally don't need to define classes to capture `args, kwargs` because Python has 'closure' that allows definition of functions to capture variables from its outer scope. And it is generally equivalent to defining extra object to store args and kwargs.
- It has equivalent feature as [lazy-string](https://pypi.org/project/lazy-string/), however, I think that my suggestion is generally better than that because it is more well-typed. [lazy-string](https://pypi.org/project/lazy-string/) is not completely typed, because it receives general type of functions as `func`, however the type of `func` and `args, kwargs` does not match. Although I think that the typing of [lazy-string](https://pypi.org/project/lazy-string/) can be improved by using `ParamSpec`, however, only a few skilled programmers in Python community are familiar with such features, and I would like to avoid such difficult to understand program if there are much concise alternatives. And even though the discussion is not for the point of making mypy or pyright happier, however, the conclusion is generally relevant to the conciseness of code and writing more easier to understand/debug code, and it is more easier to understand and debug code that takes much simple types of functions (functions that receives no argument) than functions that are too general such that it works with arbitrary arguments.

Reply:
Sounds good to me. Thanks for detailed explanation.

There's a similar API in pytorch too.
```
torch._check(a > 0, lambda: f"Need a > 0, but got {a}")
```